test/e2e/controller_test.go

```diff
@@ -0,0 +1,401 @@
+		}
+		return reason == "", nil
+	}); err != nil {
+		t.Fatalf("workspace %q never ready: %v", clusterName, err)
+	}
+
+	return createAPIBinding(t, clusterName)
+}
+
+func createAPIBinding(t *testing.T, workspaceCluster logicalcluster.Name) client.Client {
+	c := loadClient(t, workspaceCluster)
+	apiName := "catalog.kcp.dev"
+	t.Logf("creating APIBinding %q|%q", workspaceCluster, apiName)
+	if err := c.Create(context.TODO(), &apisv1alpha1.APIBinding{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: apiName,
+		},
+		Spec: apisv1alpha1.APIBindingSpec{
+			Reference: apisv1alpha1.ExportReference{
+				Workspace: &apisv1alpha1.WorkspaceExportReference{
+					Path:       parentWorkspace(t).String(),
+					ExportName: apiName,
+				},
+			},
+		},
+	}); err != nil {
+		t.Fatalf("could not create APIBinding %q|%q: %v", workspaceCluster, apiName, err)
+	}
+
+	t.Logf("waiting for APIBinding %q|%q to be bound", workspaceCluster, apiName)
+	var apiBinding apisv1alpha1.APIBinding
+	if err := wait.PollImmediate(100*time.Millisecond, wait.ForeverTestTimeout, func() (done bool, err error) {
+		fetchErr := c.Get(context.TODO(), client.ObjectKey{Name: apiName}, &apiBinding)
+		if fetchErr != nil {
+			t.Logf("failed to get APIBinding %q|%q: %v", workspaceCluster, apiName, err)
+			return false, fetchErr
+		}
+		var reason string
+		if !conditions.IsTrue(&apiBinding, apisv1alpha1.InitialBindingCompleted) {
+			condition := conditions.Get(&apiBinding, apisv1alpha1.InitialBindingCompleted)
+			if condition != nil {
+				reason = fmt.Sprintf("%q: %q", condition.Reason, condition.Message)
+			} else {
+				reason = "no condition present"
+			}
+			t.Logf("not done waiting for APIBinding %q|%q to be bound: %q", workspaceCluster, apiName, reason)
+		}
+		return conditions.IsTrue(&apiBinding, apisv1alpha1.InitialBindingCompleted), nil
+	}); err != nil {
+		t.Fatalf("APIBinding %q|%q never bound: %v", workspaceCluster, apiName, err)
+	}
+
+	return c
+}
+
+func createAPIResourceSchema(t *testing.T, c client.Client, workspaceCluster logicalcluster.Name, schema *apisv1alpha1.APIResourceSchema) error {
+	apiName := schema.GetName()
+	t.Logf("creating APIResourceSchema %q|%q", workspaceCluster, apiName)
+	if err := c.Create(context.TODO(), schema); err != nil {
+		return fmt.Errorf("could not create APIResourceSchema %q|%q: %v", workspaceCluster, apiName, err)
+	}
+
+	t.Logf("waiting for APIResourceSchema %q|%q to be found", workspaceCluster, apiName)
+	var apiResourceSchema apisv1alpha1.APIResourceSchema
+	if err := wait.PollImmediate(100*time.Millisecond, wait.ForeverTestTimeout, func() (done bool, err error) {
+		fetchErr := c.Get(context.TODO(), client.ObjectKey{Name: apiName}, &apiResourceSchema)
+		if fetchErr != nil {
+			t.Logf("failed to get APIResourceSchema %q|%q: %v", workspaceCluster, apiName, err)
+			return false, fetchErr
+		}
+		return true, nil
+	}); err != nil {
+		return fmt.Errorf("APIResourceSchema %q|%q not found: %v", workspaceCluster, apiName, err)
+	}
```

Comment:
This isn't strictly required - we know that if we created a thing, it is immediately gettable